src/librustc_metadata/tyencode.rs

```diff
@@ -407,7 +407,14 @@ pub fn enc_existential_bounds<'a,'tcx>(w: &mut Cursor<Vec<u8>>,
 
     enc_region(w, cx, bs.region_bound);
 
-    for tp in &bs.projection_bounds {
+    // Encode projection_bounds in a stable order
+    let mut projection_bounds: Vec<_> = bs.projection_bounds
+                                          .iter()
+                                          .map(|b| (b.item_name().as_str(), b))
+                                          .collect();
+    projection_bounds.sort_by_key(|&(ref name, _)| name.clone());
```

Comment:
Since sort_by_key() returns nothing, projection_bounds with be the unit value `()`. It would have to look at least like:
```rust
let mut projection_bounds = bs.projection_bounds.clone();
projection_bounds.sort_by_key(|&b| b.item_name().as_str());
```

And why didn't I do it like the above? In order to avoid that `b.item_name().as_str()` gets called more than once per item. Although, I somehow mistakenly assumed that `b.item_name().as_str()` would need to hash the string each time when looking it up in the interner, which it doesn't, so this "optimization" is even more premature than I thought `:)`

So, your version, when fixed as indicated, would actually be nicer than mine.